include/matx/core/operator_utils.h

```diff
@@ -62,6 +64,30 @@ namespace matx {
     return func(in, iter, bi, ei);
   }  
 
+  template <typename In, typename Out, typename Op, bool ConvertType = true>
+  __MATX_HOST__ __MATX_INLINE__ auto ReduceInputThrust(In &&in, Out &&out, Op &&op) {
+    typename detail::base_type_t<In> in_base = in;    
+
+    auto begin = BeginOffset{in_base};
+    auto end = EndOffset{in_base};
+
+    if constexpr (in_base.Rank() < 2 && is_tensor_view_v<In>) {
+      using value_t = typename Out::value_type;
+      const auto &iter = matx::RandomOperatorIterator<decltype(in_base), ConvertType>{in_base};
+
+      if (in_base.IsContiguous()) {
+        // the conversion is already handled for us by RandomOperatorIterator
+        thrust::reduce(
+            iter + *begin, iter + *end, op.Init(), op
```

Comment:
@cliffburdick 

So this function is responsible for dealing with the facct that `In` may not necessarily be a `tensor_t` but rather an operator of some sort. It'll get the respective offsets and construct an iterator for thrust to use, which, thankfully, does seem to be perfectly compatible. 

Consequentially it would seem the code has also become potentially simpler than it's counterpart on `main`. Thrust is now responsible for deciding whom exactly to use in CUB rather than MatX. I might be missing some context though on whether or not MatX needing to be responsible for what function to use in CUB/Thrust.

Reply:
In hindsight, I realize now that there might be issues with this approach. Does `matXBinaryOp` for example need to be utilized? I see that it has methods like `PreRun` `PostRun` etc that make use of the Executor(which I have used here)

Reply:
Hi @ZelboK since a tensor _is_ an operator, the iterator wrapper can turn any MatX operator type into an iterator. However, we do the contiguous check to allow CUB/thrust an optimization if it's a flat pointer with contiguous strides.

Regarding your second comment, the `ReduceInput` function shouldn't/doesn't need to know whether it's CUB or thrust. That's something you pass in your lambda that you give to the function. I'm saying this all without actually trying it, but if you assume the iterator type is compatible between both libraries then there may not be any changes to that code. This was how the example I pointed to previously worked (just search ReduceInput in cub.h).

I'm not sure what your question about `matxBinaryOp` is, but that's a wrapper class for any binary type and should be completely separate from this. `matxBinaryOp`, like most of our types is an operator, and you can pass it to thrust/cub and have the iterator pull from it.

Reply:
@cliffburdick I see. I should have clarified - I actually did try to follow that example you've shown me like this(but I did it incorrectly). If you're curious, here's what I did

```
auto output_ = cuda::std::get<0>(out);
        using out_t = decltype(output_);
        using value_t = typename out_t::value_type;
        using input_t = typename detail::base_type_t<OpA>;
        using output_t = typename detail::base_type_t<out_t>;
        input_t in_base = a_;
        output_t out_base = output_;
        auto op = detail::reduceOpAny<value_t>();

        auto fn = [&](input_t &&input, 
          output_t && output, 
          BeginOffset<input_t> &&begin,  
          EndOffset<input_t> &&end) { 
            return thrust::reduce(
              input + *begin, input + *end, op.Init(), op
            );
        };
        auto rv = ReduceInput(fn, out_base, in_base);
```

which brought me to my question of `matXBinaryOp`. This led to errors like

```
has no member "type"
    typedef typename thrust::iterator_system<InputIterator>::type System;   
```
where `InputIterator` was of type
```
matx::detail::matxBinaryOp<matx::detail::ConstVal<int, cuda::std::__4::array<matx::index_t, 2UL>>, matx::index_t, matx::detail::BinOp<int, matx::index_t, matx::detail::AddF<int, matx::index_t>>>
```
I should not be passing the base types though. Let me push the fix.